stdnum/de/steuernummer.py

```diff
@@ -0,0 +1,171 @@
+    ...
+InvalidLength: ...
+"""
+from stdnum.exceptions import (
+    InvalidLength,
+    InvalidFormat
+)
+from stdnum.util import clean
+
+
+def compact(number):
+    """
+    Convert the number to the minimal representation. This strips the
+    number of any valid separators ' -./,' and removes surrounding whitespace.
+    """
+    number = clean(number, ' -./,').upper().strip()
+    return number
+
+
+state_opening_character = {
+    'Baden-Württemberg': {
+        'standard': '',
+        'bund': '28'
+    },
+    'Bayern': {
+        'standard': '',
+        'bund': '9'
+    },
+    'Berlin': {
+        'standard': '',
+        'bund': '11'
+    },
+    'Brandenburg': {
+        'standard': '0',
+        'bund': '30'
+    },
+    'Bremen': {
+        'standard': '',
+        'bund': '24'
+    },
+    'Hamburg': {
+        'standard': '',
+        'bund': '22'
+    },
+    'Hessen': {
+        'standard': '0',
+        'bund': '26'
+    },
+    'Mecklenburg-Vorpommern': {
+        'standard': '0',
+        'bund': '40'
+    },
+    'Niedersachsen': {
+        'standard': '',
+        'bund': '23'
+    },
+    'Nordrhein-Westfalen': {
+        'standard': '',
+        'bund': '5'
+    },
+    'Rheinland-Pfalz': {
+        'standard': '',
+        'bund': '27'
+    },
+    'Saarland': {
+        'standard': '0',
+        'bund': '10'
+    },
+    'Sachsen': {
+        'standard': '2',
+        'bund': '32'
+    },
+    'Sachsen-Anhalt': {
+        'standard': '1',
+        'bund': '31'
+    },
+    'Schleswig-Holstein': {
+        'standard': '',
+        'bund': '21'
+    },
+    'Thüringen': {
+        'standard': '1',
+        'bund': '41'
+    }
+}
+
+
+def get_state_opening_character(state, schema):
+    """
+    Return a string of the expected openeing chracter of that state.
+    The return value will depend on the specification for that state and the
+    schema type, schema type should be:
+    standard: For old standard schema (Standardschema der Länder) or
+    (Standardschema)
+    bund: For the new federal schema (Vereinheitlichtes Bundesschema) or
+    (Bundesschema)
+    Use state name as the refernce in the wikipedia article.
+    """
+    return state_opening_character[state][schema]
+
+
+def is_number_acceptable_to_state(number, state, schema):
+    """
+    Checks if the number is acceptable for a state.
+    """
+    opening_chars = get_state_opening_character(state, schema)
+    len_opening_chars = len(opening_chars)
+    if number[:len_opening_chars] == opening_chars:
+        return number
+    raise InvalidFormat('This identifier is not acceptable for %s' % state)
+
+
+def validate(number, state=None):
+    """
+    Checks to see if the number provided is a valid tax number.
+    This checks the length and formatting.
+    """
+    clean_number = compact(number)
+    if not clean_number.isdigit():
+        raise InvalidFormat('The number contains letter.')
+    if len(clean_number) == 13:
```

Comment:
I'd add private functions like _validate_standart_schema(...) and _validate_federal_schema(...)